### ISSUE TITLE
prediction output have position shift

### ISSUE THREAD

**Issue by ylfmsn** [azavea/raster-vision#1773]
When I use the trained forest image data model to predict another image data, why does the output geojson result have a position shift?

![image](https://user-images.githubusercontent.com/23330401/231640656-c903c843-7a55-4058-860e-384868978a53.png)
![image](https://user-images.githubusercontent.com/23330401/231640723-1a2346e3-0b37-4457-b762-6a199a012e70.png)

**Comment by lewfish** [comment]
Are you saying that you trained your own model to predict forests, and this was the output? My guess is that the labels were misaligned with your imagery in the training set and the model has learned to reproduce this behavior. I would look at your training images and labels in a tool like QGIS and check the alignment.

**Comment by lewfish** [comment]
If your training labels are in a vector format like GeoJSON, you should be able to shift them using the [ShiftTransformer](https://docs.rastervision.io/en/0.20/api_reference/_generated/rastervision.core.data.vector_transformer.shift_transformer.ShiftTransformer.html#rastervision.core.data.vector_transformer.shift_transformer.ShiftTransformer) which is a type of `VectorTransformer`.

**Comment by ylfmsn** [comment]
Yes, I trained the model first, and then used the trained model to predict woodland，the screenshot below is the label and image of the training，the image data format is tif，the label data format is geojson, the coordinate system is wgs84：
![image](https://user-images.githubusercontent.com/23330401/232190312-911db3ff-f430-4d3e-8aaf-0fa8e532d30b.png)
![image](https://user-images.githubusercontent.com/23330401/232190353-42e86ac5-21c5-4682-97b5-c7045c276dfc.png)

**Comment by ylfmsn** [comment]
> When I use the trained forest image data model to predict another image data, why does the output geojson result have a position shift?
> 
> ![image](https://user-images.githubusercontent.com/23330401/231640656-c903c843-7a55-4058-860e-384868978a53.png) ![image](https://user-images.githubusercontent.com/23330401/231640723-1a2346e3-0b37-4457-b762-6a199a012e70.png)

this is the output

**Comment by ylfmsn** [comment]
> When I use the trained forest image data model to predict another image data, why does the output geojson result have a position shift?
> 
> ![image](https://user-images.githubusercontent.com/23330401/231640656-c903c843-7a55-4058-860e-384868978a53.png) ![image](https://user-images.githubusercontent.com/23330401/231640723-1a2346e3-0b37-4457-b762-6a199a012e70.png)

This is the predict output. The training dataset does not seem to be shifted, but why is the output shifted? I'm confused...

**Comment by lewfish** [comment]
Which version of RV are you using and how did you install it? Would you be able to share the model bundle and a sample image where this is happening?

**Comment by AdeelH** [comment]
@ylfmsn Is the raster prediction output (`labels.tif`) also shifted or just the GeoJSON?

**Comment by ylfmsn** [comment]
> @ylfmsn Is the raster prediction output (`labels.tif`) also shifted or just the GeoJSON?

labels.tif is also shifted

**Comment by ylfmsn** [comment]
> Which version of RV are you using and how did you install it? Would you be able to share the model bundle and a sample image where this is happening?

the raster-vision version is 0.20.2

sample image:
[sample image.zip](https://github.com/azavea/raster-vision/files/11674279/sample.image.zip)

**Comment by AdeelH** [comment]
I am not able to reproduce the misalignment issue. This is the output I get when I use your model to predict on the sample image:
![image](https://github.com/azavea/raster-vision/assets/13014700/fbc4c20c-0c97-4b53-8c42-a3c872da7cc3)

This is the code I used (from the training and prediction tutorial notebooks), with RV v0.20.2:
```py
import torch
import albumentations as A

from rastervision.core.data import PolygonVectorOutputConfig
from rastervision.pytorch_learner import SemanticSegmentationLearner
from rastervision.core.data import ClassConfig
from rastervision.pytorch_learner import SemanticSegmentationSlidingWindowGeoDataset
from rastervision.core.data import SemanticSegmentationLabels

model = torch.hub.load(
    'AdeelH/pytorch-fpn:0.3',
    'make_fpn_resnet',
    name='resnet18',
    fpn_type='panoptic',
    num_classes=2,
    fpn_channels=128,
    in_channels=3,
    out_size=(256, 256),
    pretrained=True)

learner = SemanticSegmentationLearner.from_model_bundle(
    model_bundle_uri='model/model-bundle.zip',
    output_dir='debug-1773/',
    model=model,
)

ds = SemanticSegmentationSlidingWindowGeoDataset.from_uris(
    class_config=class_config,
    image_uri='sample.tif',
    size=256,
    stride=256,
    transform=A.Resize(256, 256))

predictions = learner.predict_dataset(
    ds,
    raw_out=True,
    numpy_out=True,
    predict_kw=dict(out_shape=(256, 256)),
    progress_bar=True)

pred_labels = SemanticSegmentationLabels.from_predictions(
    ds.windows,
    predictions,
    smooth=True,
    extent=ds.scene.extent,
    num_classes=len(class_config))

pred_labels.save(
    uri='pred/',
    crs_transformer=ds.scene.raster_source.crs_transformer,
    class_config=class_config,
    discrete_output=True,
    smooth_output=False,
    vector_outputs=[PolygonVectorOutputConfig(class_id=1, uri='pred/vector_output/class-1-polygons.json')]
)
```
<hr>

That said, there does seem to be something off with the CRS in your sample TIFF. The above code throws a GDAL warning when reading from it. The warning can also be seen via `gdalinfo`:

`gdalinfo "sample.tif"`

```
Warning 1: The definition of projected CRS EPSG:2363 got from GeoTIFF keys is not the same as the one from the EPSG registry, which may cause issues during reprojection operations. Set GTIFF_SRS_SOURCE configuration option to EPSG to use official parameters (overriding the ones from GeoTIFF keys), or to GEOKEYS to use custom values from GeoTIFF keys and drop the EPSG code.
```

Comparison of the CRS portion of the outputs of `gdalinfo "sample.tif"` (right) with `gdalsrsinfo "EPSG:2363"` (left) looks like so:
![image](https://github.com/azavea/raster-vision/assets/13014700/edbf194c-b94c-4d98-ac73-876cd96ce5ec)

I think the diffs on lines 9, 28, and 31 are significant.

<hr>

Maybe you can try the suggestions in the warning and see if it helps.

>Set GTIFF_SRS_SOURCE configuration option to EPSG to use official parameters (overriding the ones from GeoTIFF keys), or to GEOKEYS to use custom values from GeoTIFF keys and drop the EPSG code.

You can do this via, e.g.:
```py
import os
os.environ['GTIFF_SRS_SOURCE'] = 'GEOKEYS'
```
or

```py
import os
os.environ['GTIFF_SRS_SOURCE'] = 'EPSG'
```

I tried both and didn't seem to notice any difference though. My predictions always seem to be aligned correctly.

**Comment by AdeelH** [comment]
@ylfmsn Note that I tested the above inside the RV docker image. It is possible that the difference is due a difference in the GDAL version. The GDAL version (`gdalinfo --version`) in the docker image is `3.5.2`. Maybe you can re-try with that version.

Also, the rasterio version is `1.3.2`.

**Comment by ylfmsn** [comment]
@AdeelH Thank you very much for your help, I think I know what the problem is, data offset has bothered me for a long time, I will check the data again, thank you again！

**Comment by ylfmsn** [comment]
> @ylfmsn Note that I tested the above inside the RV docker image. It is possible that the difference is due a difference in the GDAL version. The GDAL version (`gdalinfo --version`) in the docker image is `3.5.2`. Maybe you can re-try with that version.
> 
> Also, the rasterio version is `1.3.2`.

Why is the output geojson different from the output labels.tif coordinate system? It seems that no matter what the coordinate system of sample.tif is, the output geojson coordinate system is wgs84, but the coordinate system of label.tif can always be consistent with sample.tif...

**Comment by ylfmsn** [comment]
@AdeelH I used the coordinate system epsg:2363 of sample.tif for prediction. Why is the output geojson coordinate system wgs84? It seems that no matter what coordinate system I use for sample.tif, the coordinate system of the output geojson is wgs84. How can I set the coordinate system of the output geojson to be the same as that of the input sample.tif?

**Comment by AdeelH** [comment]
The official GeoJSON standard [only supports WGS84](https://stevage.github.io/geojson-spec/#section-4). RV complies with that standard. You may be able to use other tools to convert the GeoJSON to other coordinate systems.

Note that there is no similar restriction on the CRS of the GeoTIFF. But the input GeoJSON (for training) is also expected to be WGS84.
